keps/sig-api-machinery/2885-server-side-unknown-field-validation/README.md

```diff
@@ -0,0 +1,916 @@
+
+<!--
+This section must be completed when targeting alpha to a release.
+-->
+
+###### How can this feature be enabled / disabled in a live cluster?
+
+<!--
+Pick one of these and delete the rest.
+-->
+- [x] Feature gate (also fill in values in `kep.yaml`)
+  - Feature gate name: UnknownFieldValidation
+  - Components depending on the feature gate: kube-apiserver
+- [x] Other
+  - Describe the mechanism: query parameter
+  - Will enabling / disabling the feature require downtime of the control
+    plane? NO
+  - Will enabling / disabling the feature require downtime or reprovisioning
+    of a node? (Do not assume `Dynamic Kubelet Config` feature is enabled). NO
+
+###### Does enabling the feature change any default behavior?
+
+No, strict validation is false by default.
+
+###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?
+
+Yes, not setting the query param will ensure we are not doing strict validation
+(i.e. disable the feature)
+
+###### What happens if we reenable the feature if it was previously rolled back?
+
+No harm, requests will resume performing strict validation
+
+###### Are there any tests for feature enablement/disablement?
+
+Testing for both presenence and absence of query param.
+
+### Rollout, Upgrade and Rollback Planning
+
+This section must be completed when targeting beta to a release.
+
+N/A
+
+<!--
+###### How can a rollout or rollback fail? Can it impact already running workloads?
+
+Try to be as paranoid as possible - e.g., what if some components will restart
+mid-rollout?
+
+Be sure to consider highly-available clusters, where, for example,
+feature flags will be enabled on some API servers and not others during the
+rollout. Similarly, consider large clusters and how enablement/disablement
+will rollout across nodes.
+
+###### What specific metrics should inform a rollback?
+
+What signals should users be paying attention to when the feature is young
+that might indicate a serious problem?
+
+###### Were upgrade and rollback tested? Was the upgrade->downgrade->upgrade path tested?
+
+Describe manual testing that was done and the outcomes.
+Longer term, we may want to require automated upgrade/rollback tests, but we
+are missing a bunch of machinery and tooling and can't do that now.
+
+###### Is the rollout accompanied by any deprecations and/or removals of features, APIs, fields of API types, flags, etc.?
+
+Even if applying deprecation policies, they may still surprise some users.
+-->
+### Monitoring Requirements
+
+This section must be completed when targeting beta to a release.
+
+N/A
+<!--
+###### How can an operator determine if the feature is in use by workloads?
+
+Ideally, this should be a metric. Operations against the Kubernetes API (e.g.,
+checking if there are objects with field X set) may be a last resort. Avoid
+logs or events for this purpose.
+
+###### How can someone using this feature know that it is working for their instance?
```

Comment:
While not strictly required for alpha, this seems like a worthwhile question to be answered given how broadly it applies.  How does an end-user distinguish, "my request for strict validation resulted in zero errors" from "my request for strict validation was ignored"?

Reply:
Do you think it is sufficient to say that an end user can cross-reference the results of client-side validation (i.e. `kubectl --validate=true`) with server-side (i.e. hitting the api-server with the query param)? Meaning if client-side validation fails, but server-side validation does not indicate a failure, then the feature is disabled, but if both CS and SS succeed, then it resulted in no errors.

Is there some other, more definitive way for an end-user to know what feature-flags are set on the API server?

Reply:
Perhaps adding a nonsense field in objectmeta, where no resource can accept extra fields, and doing a dry-run.  If an error is returned, the feature is on?

As a user, I'm trying to work out how long until I would wait until I trusted that this queryparam would do what I wanted.